diff --git a/mrbgems/picoruby-machine/include/machine.h b/mrbgems/picoruby-machine/include/machine.h
--- a/mrbgems/picoruby-machine/include/machine.h
+++ b/mrbgems/picoruby-machine/include/machine.h
@@ -68,1 +68,2 @@
 void Machine_reboot(void);
+void Machine_bootsel(void);
diff --git a/mrbgems/picoruby-machine/mrblib/machine.rb b/mrbgems/picoruby-machine/mrblib/machine.rb
--- a/mrbgems/picoruby-machine/mrblib/machine.rb
+++ b/mrbgems/picoruby-machine/mrblib/machine.rb
@@ -22,0 +23,4 @@
+  def self.bootsel
+    self._bootsel
+  end
+
diff --git a/mrbgems/picoruby-machine/ports/rp2040/machine.c b/mrbgems/picoruby-machine/ports/rp2040/machine.c
--- a/mrbgems/picoruby-machine/ports/rp2040/machine.c
+++ b/mrbgems/picoruby-machine/ports/rp2040/machine.c
@@ -10,0 +11,1 @@
+#include "pico/bootrom.h"
@@ -42,0 +44,6 @@
+void
+Machine_bootsel(void)
+{
+  rom_reset_usb_boot(0, 0);
+}
+
diff --git a/mrbgems/picoruby-machine/src/mruby/machine.c b/mrbgems/picoruby-machine/src/mruby/machine.c
--- a/mrbgems/picoruby-machine/src/mruby/machine.c
+++ b/mrbgems/picoruby-machine/src/mruby/machine.c
@@ -496,0 +497,11 @@
+static mrb_value
+mrb_s__bootsel(mrb_state *mrb, mrb_value self)
+{
+#if !defined(PICORB_PLATFORM_POSIX)
+  Machine_bootsel();
+#else
+  mrb_raise(mrb, E_NOTIMP_ERROR, "Machine._bootsel is not available on this platform");
+#endif
+  return mrb_nil_value();
+}
+
@@ -529,0 +541,1 @@
+  mrb_define_class_method_id(mrb, module_Machine, MRB_SYM(_bootsel), mrb_s__bootsel, MRB_ARGS_NONE());
diff --git a/mrbgems/picoruby-machine/src/mrubyc/machine.c b/mrbgems/picoruby-machine/src/mrubyc/machine.c
--- a/mrbgems/picoruby-machine/src/mrubyc/machine.c
+++ b/mrbgems/picoruby-machine/src/mrubyc/machine.c
@@ -316,0 +317,11 @@
+static void
+c_Machine__bootsel(mrbc_vm *vm, mrbc_value *v, int argc)
+{
+#if !defined(PICORB_PLATFORM_POSIX)
+  Machine_bootsel();
+#else
+  mrbc_raise(vm, MRBC_CLASS(RuntimeError), "Machine.bootsel is not available on this platform.");
+#endif
+  SET_NIL_RETURN();
+}
+
@@ -546,0 +558,1 @@
+  mrbc_define_method(vm, module_Machine, "_bootsel", c_Machine__bootsel);
diff --git a/mrbgems/picoruby-shell/shell_executables/_path.txt b/mrbgems/picoruby-shell/shell_executables/_path.txt
--- a/mrbgems/picoruby-shell/shell_executables/_path.txt
+++ b/mrbgems/picoruby-shell/shell_executables/_path.txt
@@ -17,0 +18,1 @@
+/bin/bootsel
diff --git a/mrbgems/picoruby-shell/shell_executables/bootsel.rb b/mrbgems/picoruby-shell/shell_executables/bootsel.rb
new file mode 100644
--- /dev/null
+++ b/mrbgems/picoruby-shell/shell_executables/bootsel.rb
@@ -0,0 +1 @@
+Machine.bootsel
